intake_esm/core.py

```diff
@@ -1,242 +1,475 @@
         else:
-            return input_collection
-
-    def _build_collection(self, overwrite_existing):
-        """ Build a collection defined in a YAML input file or a dictionary of nested dictionaries"""
-        name = self.input_collection['name']
-        collection_type = self.input_collection['collection_type']
-        if name not in self.collections or overwrite_existing:
-            cc = ESMMetadataStoreCatalog.collection_types[collection_type]
-            cc = cc(self.input_collection, fs=self.fs)
-            cc.build()
-            self.collections = _get_built_collections()
-        self.open_collection(name)
-
-    def _get_s3_connection_info(self):
-        try:
-            if 'requester_pays' not in self.storage_options:
-                self.storage_options['requester_pays'] = True
-            self.fs = s3fs.S3FileSystem(**self.storage_options)
-        except Exception as exc:
-            raise exc
-
-    def open_collection(self, collection_name):
-        """ Open an ESM collection """
-        self.ds = _open_collection(collection_name)
-        self.collection_name = self.ds.attrs['name']
-        self.collection_type = self.ds.attrs['collection_type']
+            groups = self.df.groupby(self.df.columns.tolist())
+        print(
+            f"""--> The keys in the returned dictionary of datasets are constructed as follows:\n\t'{".".join(groupby_attrs)}'"""
+        )
+        print(f'\n--> There will be {len(groups)} group(s)')
+
+        dsets = [
+            _load_group_dataset(
+                key,
+                df,
+                self._col_data,
+                agg_columns,
+                aggregation_dict,
+                path_column_name,
+                variable_column_name,
+                use_format_column,
+                mapper_dict,
+                self.zarr_kwargs,
+                self.cdf_kwargs,
+            )
+            for key, df in groups
+        ]
+
+        dsets = dask.compute(*dsets)
+        del mapper_dict
+
+        self._ds = {dset[0]: dset[1] for dset in dsets}
+
+
+def _unique(df, columns):
+    if isinstance(columns, str):
+        columns = [columns]
+    if not columns:
+        columns = df.columns
+
+    info = {}
+    for col in columns:
+        uniques = df[col].unique().tolist()
+        info[col] = {'count': len(uniques), 'values': uniques}
+    return info
+
+
+@dask.delayed
+def _load_group_dataset(
+    key,
+    df,
+    col_data,
+    agg_columns,
+    aggregation_dict,
+    path_column_name,
+    variable_column_name,
+    use_format_column,
+    mapper_dict,
+    zarr_kwargs,
+    cdf_kwargs,
+):
+
+    aggregation_dict = copy.deepcopy(aggregation_dict)
+    agg_columns = agg_columns.copy()
+    drop_cols = []
+    for col in agg_columns:
+        if df[col].isnull().all():
+            drop_cols.append(col)
+            del aggregation_dict[col]
+
+    agg_columns = list(filter(lambda x: x not in drop_cols, agg_columns))
```

Comment:
@rabernat, if you have a minute, have a look at this code here. 

This enables us to handle situations where the aggregation operation doesn't apply to the data returned by a query. For instance, `fx` data doesn't have `time`, so the `time_range` column is all NaNs. We therefore identify this situation and skip that aggregation. The also enables us to handle DCPP data, which has the `forecast_init_year` dimension.

Reply:
It seems like this logic could break if `col` is a column in the CSV but not a variable in the file. (This is not guaranteed.)

Wouldn't a simpler check for whether to skip the aggregation just be whether the number of items to aggregate is 1?

Reply:
> It seems like this logic could break if col is a column in the CSV but not a variable in the file. (This is not guaranteed.) 

Could you expand on this? For instance, `time_range` is not a variable in the file, but having 

```json
   {
        "type": "join_existing",
        "attribute_name": "time_range",
        "options": { "dim": "time" }
      },
``` 

works.

Reply:
@rabernat, I am not sure I follow your logic. The aggregation columns should be consistent with the datasets in the files by definition. Can you provide an example situation that you think might be problematic?

I was thinking we might add a check for some, but not all NaNs:
```python
if df[col].isnull().all():
   drop_cols.append(col)
   del aggregation_dict[col]
elif df[col].isnull().any():
   raise Error
```
Though this could/should be applied at the collection generation step.

I think we could skip aggregations where N = 0, return, perhaps an empty dataset. It's not clear to me that this is preferable, but maybe I am missing something.

Reply:
Yup, I was off base here. I was misreading `df[col]` as `ds[col]` where `ds` is an xarray dataset.

Comment withdrawn. At this point, let's just ship this and clean up any bugs we find post hackathon.

Reply:
Whew! Agreed! 

We have a test suite running that assembles random queries and tracks failures. We've now done 100s of tests and it seems that failures are due to underlying problems with the data (missing variables for some ensemble members, broken links, etc.). Ultimately we might add logic to pre-empt failures in these cases, but for now it'll be up to the user to workaround them by increasing the specificity of queries. 

Good news is, however, most tests pass!